src/main/java/hudson/plugins/git/extensions/impl/BrowserGuesser.java

```diff
@@ -0,0 +1,27 @@
+package hudson.plugins.git.extensions.impl;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.plugins.git.browser.BitbucketWeb;
+import hudson.plugins.git.browser.GitLab;
+import hudson.plugins.git.browser.GitRepositoryBrowser;
+import hudson.plugins.git.browser.GithubWeb;
+import jenkins.model.Jenkins;
+
+public class BrowserGuesser implements ExtensionPoint {
+    public GitRepositoryBrowser guessBrowser(String url){
```

Comment:
I think you should have an abstract class with an abstract `guessBrowser` method, and then a non-abstract class that is registered using the `@Extension` annotation, extends the abstract class, and recognizes the known URLs. This way, if other plugins define their own classes that extend the abstract class, they won't inherit the implementation that recognizes `https://bitbucket.org/` etc.

Or you could even define three non-abstract classes in git-plugin; one each for BitbucketWeb, GitLab and GithubWeb. But I don't know whether that would really be any more flexible.

Reply:
I feel it's better to have an abstract class and extend it for further implementation.

Reply:
I meant, if you have an abstract class BrowserGuesser, then you can define static inner classes BitbucketWeb.BrowserGuesserImpl, GitLab.BrowserGuesserImpl, and GithubWeb.BrowserGuesserImpl, each of which extends BrowserGuesser. But I don't know whether that would really be better than having one class BrowserGuesserImpl recognize all URL formats. On the contrary, splitting the implementation like that could make it more difficult to verify that the implementations don't try to recognize the same URL.

Reply:
Not sure which one would be a better implementation. We can proceed with either of them.